src/main/java/duke/logic/LCommandParser.java

```diff
@@ -0,0 +1,126 @@
+     * @param storage The storage logic that allows the command parser to write the task list data to it.
+     * @param taskList The list of tasks.
+     */
+    public LCommandParser(String input, TaskList taskList, LStorage storage, Ui ui) {
+        if (input == null || input.equals("")) {
+            throw new DukeException(EMPTY_INPUT_MESSAGE);
+        }
+        String[] inputArr = input.split(" ", 2);
+        LCommandsEnum commandEnum;
+        try {
+            commandEnum = LCommandsEnum.valueOf(inputArr[0].toUpperCase());
+        } catch (IllegalArgumentException e) {
+            throw new DukeException(INVALID_COMMAND);
+        }
+        this.willExit = false;
+        switch (commandEnum) {
+        case BYE:
+            ui.sayGoodBye();
+            willExit = true;
+            break;
+        case LIST:
+            ui.printAllTasks(taskList.getAllTasks(), taskList.size());
+            break;
+        case UPCOMING:
+            ui.printUpcomingTasks(taskList.getUpcomingTasks(), taskList.size());
+            break;
+        case HELP:
+            try {
+                ui.displayHelp(inputArr.length == 1 ? null : LCommandsEnum.valueOf(inputArr[1].toUpperCase()));
+            } catch (IllegalArgumentException e) {
+                throw new DukeException(INVALID_COMMAND);
+            }
+            break;
+        default:
+            if (inputArr.length < 2) {
+                throw new DukeException(TOO_LITTLE_ARGUMENTS_MESSAGE);
+            }
+            switch (commandEnum) {
+            case FIND:
+                ui.printTasksContaining(inputArr[1],
+                        taskList.getTasksContaining(inputArr[1]), taskList.size());
+                break;
+            case TODO: // fallthrough intended
+            case EVENT: // fallthrough intended
+            case DEADLINE:
+                ui.addTaskMessage(addTask(commandEnum.name(), inputArr[1], taskList), taskList.size());
+                break;
+            default:
+                int taskNumber;
+                try {
+                    taskNumber = Integer.parseInt(inputArr[1]);
+                } catch (NumberFormatException e) {
+                    throw new DukeException(INVALID_NUMBER_MESSAGE);
+                }
+                switch (commandEnum) {
+                case DONE:
+                    if (!taskList.markAsDone(taskNumber)) { // task already marked as done
+                        throw new DukeException("You have already marked this task (%s) as done",
+                                taskList.getTask(taskNumber));
+                    }
+                    ui.markAsDoneMessage(taskList.getTask(taskNumber));
+                    break;
+                case DELETE:
+                    ui.removeTaskMessage(taskList.removeTask(taskNumber), taskList.size());
+                    break;
+                default:
+                    throw new DukeException(INVALID_COMMAND);
+                }
```

Comment:
Good mentioned of intended fallthrough, just curious if this can be done in a single switch statement rather than nested switch statement